### ISSUE TITLE
Restarting a WebJob with a ServiceBusTrigger and AutoComplete=true can lead to data loss.

### ISSUE THREAD

**Issue by alxbl** [Azure/azure-webjobs-sdk#1573]
(This was initially opened with  the Storage SDK as https://github.com/Azure/azure-storage-net/issues/539, I have adapted the title to fit the real cause of the issue) 

I was asked to open it here for cross-referencing purposes.

# Restarting a WebJob with a `ServiceBusTrigger` and `AutoComplete=true` can lead to data loss.

It looks like when a WebJob that writes to blob storage is restarted, there is a race condition
where topic messages can be marked as completed (We use `AutoComplete = true`) but the blob isn't actually present in the storage.

## Repro

This seems to happen as soon as the WebJob needs to restart, so all the following lead to repro:

- Application Settings Changes (On Save)
- Scale Up (e.g. S1 -> S3)
- Scale Out (e.g. 1 instance -> 10 instances)
- Scale In (e.g. 10 instances -> 1 instance)
- Scale Down (e.g. S3 -> S1)
- Manual WebJob Restart (Restart and/or Stop then Start)

This doesn't seem to happen when restarting the WebApp that is hosting the WebJob. I believe this is normal because the WebJob is running on the Service Plan and not on the WebApp itself 
(Correct me if I'm wrong.)

I wrote [a sample application][2] that does away with all of our business logic and narrows it down to one of the Azure SDKs (or my ability to C#) for code review or if someone would like to try to reproduce the issue on their end.

[2]: https://github.com/alxbl/ServiceBus.MessageLoss

The best way to reproduce this issue is to play around with the application settings, scale and to manually restart the webjob while it is processing messages.

> For an average test run, I see roughly 2000-5000 blobs out of ~100K that would simply not be in blob storage, but also not be in any dead letter queue or elsewhere. This varies depending on how many times I trigger web job restarts as mentioned above.

> Reproducing can be a little time consuming, so what I usually do is setup an S3 with 10 instances of the webjob and change the application settings/restart a few times while I watch the topic queue length go down in ServiceBusExplorer. When the message count is low enough, I'll scale down to S1 1 instance until the queues hit 0 and stop the job before running the validation script (see pseudocode below)

## Configuration

**Storage Account**

- Performance = `Standard`
- Secure Transfer = `Disabled`
- Replication = `LRS`

**WebJob**

- Topic:
    - Enabled Batched Messages = `true`
    - Enabled Partitioning = `true`
- Subscription:
    - Enabled Batched Messages = `true`
    - Deadletter on filter evaluation errors = `true`
- Max Concurrent Calls = `32`
- Prefetch Count = `64`
- Auto-Renew Timeout = `120 seconds`
- Always On (Web App) = `On`

## Sample App Pseudo-Code

**WebJob**

1. Call the destination container `numbers`.
1. Let there be messages `1, 2, ..., N` in the topic `topic/sub`. `N` is the expected number of blobs.
1. For message `m` in `topic/sub`,
    1. Create a blob named `m` with arbitrary content in the `numbers` container. 
       (`await b.UploadTextAsync(m.ToString(), 'true')`)
    1. Mark message `m` as complete (`<auto-complete>`)

**Validation**

1. Retrieve all the blobs in `numbers`, call the collection `blobs`.
1. Assert  `blobs.Count` == `N`.
1. Let `expected` be the collection of numbers `1..N`.
1. Assert `expected.Except(blobs).Count == 0`.
1. If any, print `expected.Except(blobs)`.

(Assertions fail and a list of blobs is printed.)

## What we've tried so far

- Stripped away all of the business logic to rule out the application.
- Made sure cancellation tokens are passed all the way down to `CloudBlockContainer.UploadTextAsync()`.
- Tried switching `async` calls to their `sync` version.
- Upgraded to Storage SDK 8.4.0.
- Upgraded to latest WebJobs and ServiceBus SDKs.

## What we're planning to try in parallel with this issue

- Manual message completion to try and determine if `AutoComplete` is the issue
- Use REST API directly via `HttpClient` to try and rule out the Storage SDK.
- Hacky code that attempts to verify that the blob is written before returning out of the `Function` (I would rather not resort to this in our production deployments)

## Impact

As you would guess, it's a bit unnerving to deploy to production or even enable autoscale / change settings in a production environment when it carries a risk of losing production data in the process.

## Workaround

As mentioned in the linked issue, we've found that `AutoComplete = false` prevents this issue from happening, and we've resorted to using that right now.

## Actual Behavior

Messages are marked as completed before actual processing succeeds, so a shutdown will result in lost messages

## Expected Behavior

The messages should only be completed if the handler executes successfully (returns and did not throw.)

## Related Information

As I mentioned in the referenced issue:

This makes me think that either the AutoComplete documentation is [wrong][1] or there is a bug in the WebJobs.ServiceBus library (emphasis mine):

> Gets or sets a value that indicates whether the message-pump should call `CompleteAsync(String)` or `CompleteAsync(String)` on messages after the callback has **completed processing**.

> `true` to complete the message processing automatically on **successful execution of the operation**; otherwise, false.

(Notice that one sentence reads "successful execution", the other simply says "completed processing")

[1]: https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.servicebus.messagehandleroptions.autocomplete?view=azure-dotnet

To summarize, this is clearly not a Storage SDK bug (Sorry!), but might be a bug/documentation typo in the WebJobs.ServiceBus SDK.

----

Maybe the documentation should clarify what the purpose of AutoComplete is and if this is the expected behavior, clearly mention so.

Cheers,
Alex

**Comment by paulbatum** [comment]
@alxbl Thank you for the extremely thorough bug report. While I'm not a service bus expert, I do not think this is a documentation issue.

**Comment by paulbatum** [comment]
We're in the middle of looking at a few different service bus bugs that seem to relate to completion. I can't promise we'll fix this one immediately but we're at least going to do a first-pass investigation.

**Comment by alxbl** [comment]
No worries, we've rewritten the code to always explicitly complete, so there's no time pressure on this on our part. Thanks for looking into it :)

**Comment by alxbl** [comment]
@mathewc I can give it a shot sometime this weekend. Let me know when the build is available.

**Comment by mathewc** [comment]
Instructions for our myget feed are here: https://github.com/Azure/azure-webjobs-sdk/wiki/%22Nightly%22-Builds

Build **2.2.0-beta1-11248** of Microsoft.Azure.WebJobs.ServiceBus has the fix.

**Comment by alxbl** [comment]
Good morning,

I had a bit of time to try and repro this morning. I did a run with 100K messages and a S3/10 instances. I did about 10 restarts and scaled down to S3/1 for the last 2K messages.

The run had 160 messages missing, which feels like an improvement from last time... but it's been many months since I last worked on this particular project, and have since moved to a different team so it's difficult to be certain.

```
```
Unfortunately, with my current repro code I don't have much more information than that to give. I've attached my packages.config for the WebJob project, in hope that it might prove helpful.  In particular:

```
<package id="Microsoft.Azure.WebJobs" version="2.2.0-beta1-11248" targetFramework="net462" />                  
<package id="Microsoft.Azure.WebJobs.Core" version="2.2.0-beta1-11248" targetFramework="net462" />             
<package id="Microsoft.Azure.WebJobs.ServiceBus" version="2.2.0-beta1-11248" targetFramework="net462" /> 
```

The code I use to repro is available in the issue description, all that is missing would be a powershell script that automates the start/restart of the webjob in order to fully automate the test+validation run.

Let me know if I can do anything else to help.

[packages.txt](https://github.com/Azure/azure-webjobs-sdk/files/1777597/packages.txt)

**Comment by davidhjones** [comment]
Whats the current status of this Issue?

**Comment by alxbl** [comment]
As far as I'm aware it should still be possible to reproduce this issue following the repro steps provided above with the sample code. Unfortunately, I am no longer working in an Azure environment and do not have the resources to re-test. We had ended up using the workaround I described (manually completing all messages and disabling auto-complete) and it seemed to have fixed the problems. I no longer work there though so I'm not sure if that is still the case.

Cheers,
